cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java

```diff
@@ -276,15 +298,157 @@ public void dropIndex() throws IOException {
+   */
   public MetadataChange apply(MetadataMutation mutation, MutationOptions options) throws IOException {
+    try {
+      TransactionRunner runner = dbClient.readWriteTransaction();
+      return runner.run(transaction -> applyAndBufferMutation(transaction, mutation));
+    } catch (SpannerException e) {
+      throw new IOException("Error applying mutation to Spanner for entity: " + mutation.getEntity(), e);
+    }
+  }
+
+  /**
+   * Processes and buffers a single metadata mutation within a Spanner transaction.
+   *
+   * @param transaction The active Spanner {@link TransactionContext}.
+   * @param mutation    The {@link MetadataMutation} to apply.
+   * @return The resulting {@link MetadataChange}.
+   */
+  private MetadataChange applyAndBufferMutation(TransactionContext transaction, MetadataMutation mutation)
+    throws IOException{
+    MetadataEntity entity = mutation.getEntity();
+    VersionedMetadata before = readFromSpanner(entity, transaction);
+    assert before != null;
+    SpannerChangeRequest intermediary = applyMutation(before, mutation);
+    bufferMutations(transaction, intermediary.getMutation());
+    return intermediary.getChange();
+  }
+
+  /**
+   * Validates and buffers a list of mutations to be sent when the transaction commits.
+   */
+  private void bufferMutations(TransactionContext transaction, List<Mutation> mutations) {
+    for (Mutation mutation : mutations) {
+      validateMutationOperation(mutation.getOperation());
+      transaction.buffer(mutation);
+    }
+  }
+
+  /**
+   * Validates if the given {@link Mutation.Op} is a supported operation.
+   *
+   * @param operation The Spanner mutation operation to validate.
+   * @throws IllegalArgumentException if the operation is not one of the supported types
+   *                                  (INSERT, UPDATE, INSERT_OR_UPDATE, DELETE).
+   */
+  private void validateMutationOperation(Mutation.Op operation) {
+    if (!SUPPORTED_MUTATION_OPS.contains(operation)) {
+      throw new IllegalArgumentException("Unsupported Spanner Mutation operation: " + operation);
+    }
+  }
+
+  /**
+   * Creates a Spanner request that corresponds to the given mutation, along with the change
+   * effected by this mutation. The request must be executed by the caller.
+   *
+   * @param before   the metadata for the mutation's entity before the change
+   * @param mutation the mutation to apply
+   * @return a Spanner request to be executed, and the change caused by the mutation.
+   */
+  private SpannerChangeRequest applyMutation(VersionedMetadata before, MetadataMutation mutation) throws IOException {
     throw new IOException("NOT IMPLEMENTED");
   }
 
+  /**
+   * Reads the existing metadata for an entity from the table.
+   *
+   * @return existing metadata along with its version in the table, or an empty metadata with null version.
+   */
+  @VisibleForTesting
+  VersionedMetadata readFromSpanner(MetadataEntity entity, ReadContext transaction) throws
```

Comment:
nit: use better naming to describe the purpose of the function

Can be renamed as `getVersionedMetadata()` or `readVersionedMetadata()`